homeassistant/components/synology_dsm/__init__.py

```diff
@@ -144,30 +246,233 @@ async def async_setup(self):
             device_token=self._entry.data.get("device_token"),
         )
 
+        await self._should_fetch_api()
+
         await self._hass.async_add_executor_job(self._fetch_device_configuration)
-        await self.update()
+        await self.async_update()
 
         self._unsub_dispatcher = async_track_time_interval(
             self._hass,
-            self.update,
+            self.async_update,
             timedelta(
                 minutes=self._entry.options.get(
                     CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
                 )
             ),
         )
 
+    async def _should_fetch_api(self):
+        """Determine if we should fetch each API, if one entity needs it."""
+        entity_reg = await self._hass.helpers.entity_registry.async_get_registry()
```

Comment:
Ideally we register listeners and then we can determine which data to fetch by which entities are listening for it.  This avoids the need to load the whole entity registry which is an expensive operation.

Reply:
Yeah, but at HA boot I'll still need to check this right ?


Is there a way to retreive only entities of a config entry ?

Like combining
```python
        entity_reg = await self._hass.helpers.entity_registry.async_get_registry()
        entity_entries = entity_registry.async_entries_for_config_entry(
            entity_reg, self._entry.entry_id
        )
```
So it will reduce the cost of it :wink:

Reply:
You could do the full api poll without security (since you know which ones to create without polling the api as its a static list).  Then in `async_added_to_hass`  you could do something like

```
def async_added_to_hass(self):
    self.async_on_remove(self._synoapi.subscribe("security", self.async_write_ha_state))
```

Then implement a subscribe function in `SynoAPI` which keeps track of each sensor type and only does an update for those types if there are any listeners.  Then call back the listeners after an update.

Reply:
Yes but, this solution will work for `security` only because there is one  entity, but if there is more ?

I am planning to add `SurveillanceStation` and `DownloadStation`.

Can we let it like this ?

I have work already on ProtoThis/python-synology#53, then `synology` will be merged in `synology_dsm`.

Reply:
> Yes but, this solution will work for `security` only because there is one entity, but if there is more ?
> 
> I am planning to add `SurveillanceStation` and `DownloadStation`.

Seems like it would be fine to poll all the data on the initial poll since it is just once and then subsequent calls would only poll what is needed.

> 
> Can we let it like this ?
> 

I'm not sure what you are asking.

> I have work already on [ProtoThis/python-synology#53](https://github.com/ProtoThis/python-synology/issues/53), then `synology` will be merged in `synology_dsm`.

Reply:
I mean to leave it like that, for now, because this solution is working for all entities and all APIs no matter how many entities are present.

An exemple: if someone adds the integration only for `security` and `camera` (to come with `SurveillanceStation`) it should not fetch `storage` and `information`.

I'm currently trying your proposed `subscribe` way :blush:

Reply:
I think I found a way with `async_added_to_hass`

Reply:
Yeay, finally got a clean solution, see "Review: subscribe to fetching, unsubscribe on async_on_remove" commit :blush: 

Thanks of lot for beeing so patient with me :sweat_smile:

Reply:
So it fetches all at first, then enabled entities only.

Reply:
Our update coordinator helper has this feature builtin already. Why not use that, one coordinator per api endpoint?

Reply:
@MartinHjelmare 
You mean like the `light_coordinator` or `group_coordinator` from `hue/light.py` ?

I don't think I can since DSM library does not provide update method per API, but one for all:

https://github.com/ProtoThis/python-synology/blob/1286f4aaa2fae0ed13d96c350812a47683fc2476/synology_dsm/synology_dsm.py#L271-L287

Seen this doc : https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities

I should definitely change this behaviour (that was planned later).